### PR TITLE
KAFKA-4023: add thread/task id for logging prefix

### PR DESCRIPTION
**Pull request by bbejeck**

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java [hunk 1]
```diff
@@ -49,12 +49,17 @@
...
     private final Producer<byte[], byte[]> producer;
     private final Map<TopicPartition, Long> offsets;
+    private String streamTaskId = null;
 
     public RecordCollector(Producer<byte[], byte[]> producer) {
         this.producer = producer;
         this.offsets = new HashMap<>();
     }
 
+    void setStreamTask(String streamTaskId) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java]: Could we just pass in the taskId as part of the constructor? And for unit tests we can simply set a dummy taskId here.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java [hunk 2]
```diff
@@ -81,7 +86,8 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
                     TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
                     offsets.put(tp, metadata.offset());
                 } else {
-                    log.error("Error sending record to topic {}", topic, exception);
+                    String prefix = (streamTaskId == null) ? "" : "stream-task " + streamTaskId + " ";
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java]: To be consistent with other classes, should this be `"task [{taskId}] "?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 3]
```diff
@@ -106,8 +106,9 @@ public StreamTask(TaskId id,
 
         // create the record recordCollector that maintains the produced offsets
         this.recordCollector = new RecordCollector(producer);
+        this.recordCollector.setStreamTask(id().toString());
 
-        log.info("Creating restoration consumer client for stream task #" + id());
+        log.info("stream-task[{}] Creating restoration consumer client", id());
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: nit: add a space after `stream-task` to be consistent with other classes.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java [hunk 4]
```diff
@@ -149,7 +149,7 @@ public void configure(Map<String, ?> configs) {
                     (String) configs.get(StreamsConfig.ZOOKEEPER_CONNECT_CONFIG),
                     configs.containsKey(StreamsConfig.REPLICATION_FACTOR_CONFIG) ? (Integer) configs.get(StreamsConfig.REPLICATION_FACTOR_CONFIG) : 1);
         } else {
-            log.info("Config '{}' isn't supplied and hence no internal topics will be created.", StreamsConfig.ZOOKEEPER_CONNECT_CONFIG);
+            log.info("stream-thread {} Config '{}' isn't supplied and hence no internal topics will be created.",  streamThread.getName(), StreamsConfig.ZOOKEEPER_CONNECT_CONFIG);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java]: nit: `stream-thread [{}]`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 5]
```diff
@@ -220,7 +220,7 @@ public void run() {
         } catch (Exception e) {
             // we have caught all Kafka related exceptions, and other runtime exceptions
             // should be due to user application errors
-            log.error("Streams application error during processing in thread [" + this.getName() + "]: ", e);
+            log.error("stream-thread [" + this.getName() + "] Streams application error during processing: ", e);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: This is inefficient, could we use formatting like above?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 6]
```diff
@@ -258,22 +258,22 @@ private void shutdown() {
         try {
             producer.close();
         } catch (Throwable e) {
-            log.error("Failed to close producer in thread [" + this.getName() + "]: ", e);
+            log.error("stream-thread [" + this.getName() + "] Failed to close producer: ", e);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ditto above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 7]
```diff
@@ -258,22 +258,22 @@ private void shutdown() {
...
         }
         try {
             consumer.close();
         } catch (Throwable e) {
-            log.error("Failed to close consumer in thread [" + this.getName() + "]: ", e);
+            log.error("stream-thread [" + this.getName() + "] Failed to close consumer: ", e);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ditto above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 8]
```diff
@@ -258,22 +258,22 @@ private void shutdown() {
...
         }
         try {
             restoreConsumer.close();
         } catch (Throwable e) {
-            log.error("Failed to close restore consumer in thread [" + this.getName() + "]: ", e);
+            log.error("stream-thread [" + this.getName() + "] Failed to close restore consumer: ", e);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ditto above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 9]
```diff
@@ -258,22 +258,22 @@ private void shutdown() {
...
         }
 
         removeStreamTasks();
 
-        log.info("Stream thread shutdown complete [" + this.getName() + "]");
+        log.info("stream-thread [" + this.getName() + "] Stream thread shutdown complete");
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ditto above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 10]
```diff
@@ -317,7 +317,7 @@ private void runLoop() {
                 lastPoll = time.milliseconds();
 
                 if (rebalanceException != null)
-                    throw new StreamsException("Failed to rebalance", rebalanceException);
+                    throw new StreamsException("stream-thread [" + this.getName() + "]Failed to rebalance", rebalanceException);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Use `String.format`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 11]
```diff
@@ -406,8 +406,8 @@ private void maybeUpdateStandbyTasks() {
                     StandbyTask task = standbyTasksByPartition.get(partition);
 
                     if (task == null) {
-                        log.error("missing standby task for partition {}", partition);
-                        throw new StreamsException("missing standby task for partition " + partition);
+                        log.error("stream-thread [{}]  missing standby task for partition {} ", this.getName(), partition);
+                        throw new StreamsException("stream-thread [" + this.getName() + "] missing standby task for partition " + partition);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ditto above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 12]
```diff
@@ -437,7 +437,7 @@ private void maybePunctuate(StreamTask task) {
                 sensors.punctuateTimeSensor.record(computeLatency());
 
         } catch (KafkaException e) {
-            log.error("Failed to punctuate active task #" + task.id() + " in thread [" + this.getName() + "]: ", e);
+            log.error("stream-thread [" + this.getName() + "] Failed to punctuate active task #" + task.id(), e);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ditto above.

**Comment by guozhangwang** [comment]
Two meta comment:
1. We'd prefer `String.format` for exception message string and `("{}", value)` for logging message strings since they are more efficient. This may not be a big deal for those one-time message, but for some logging entries in `StreamTask` that can be a huge performance difference, see git commit #2788f2dc73dd62c35c87842d02d23a0ed6a62e97.
2. Could we add prefix in `TaskAssignor` as well?

**Comment by bbejeck** [comment]
@guozhangwang 
Thanks for the review.  

Here's a summary of changes based on the review comments:
1.  Change Exception String messages to use String.format
2.  Change Logging messages to use ("{}", value)
3.  All task logging uses consistent prefix 
4.  RecordCollector, TaskAssignor take stream-thread id as constructor arg dummy value (null) used in unit tests.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignorTest.java [hunk 13]
```diff
@@ -46,7 +46,7 @@ public void testAssignWithoutStandby() {
 
         // # of clients and # of tasks are equal.
         tasks = mkSet(0, 1, 2, 3, 4, 5);
-        assignments = TaskAssignor.assign(states, tasks, 0);
+        assignments = TaskAssignor.assign(states, tasks, 0, null);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignorTest.java]: Would this cause NPE if `INFO` logging is turned on for `TaskAssignor`?

Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignorTest.java]: @guozhangwang 

I thought the SLF4J formatting handled null values, but to be sure I tried running the test with `INFO` turned on - resulted in:  `INFO stream-thread [null]  ....`  no NPE but not very helpful information, maybe I should still set a value for more meaningful information ? Also do the same for the other tests with dummy null values ?

**Comment by bbejeck** [comment]
@guozhangwang 

Updated tests to use "test_class_name-test_method" for stream-thread-ids passed as constructor parameters.

**Comment by guozhangwang** [comment]
LGTM and merged to trunk.

**Pull request closed without merging**